File: python/neetcode/backtracking/combination_sum.py
"""
Given an array of distinct integers candidates and a target integer target,
return a list of all unique combinations of candidates where the chosen numbers sum to target.
You may return the combinations in any order.

The same number may be chosen from candidates an unlimited number of times.
Two combinations are unique if the frequency of at least one of the chosen numbers is different.

It is guaranteed that the number of unique combinations that sum up to target is less than 150 combinations for the given input.

# Solution
Backtracking

"""

def combination_sum(candidates: list[int], target: int) -> list[list[int]]:
    ans = []
    n_candidates = len(candidates)
    
    def dfs(i: int, current: list[int], total: int) -> None:
        if total == target:
            ans.append(current.copy())
            return
        if i >= n_candidates or total > target:
            return
        
        current.append(candidates[i])
        dfs(i, current, total + candidates[i])
        current.pop()
        dfs(i + 1, current, total)
        
    dfs(0, [], 0)
    return ans


if __name__ == "__main__":
    assert combination_sum([2,3,6,7], 7) == [[2,2,3], [7]]
    assert combination_sum([2,3,5], 8) == [[2,2,2,2], [2,3,3], [3,5]]
    assert combination_sum([2], 1) == []
    

File: java/neetcode/src/main/java/org/example/backtracking/CombinationSum.java
package org.example.backtracking;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class CombinationSum {

    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        var context = new Context(candidates, target);
        var stack = new Stack<Integer>();
        dfs(0, stack, 0, context);
        return context.getAns();
    }

    private static void dfs(int i, Stack<Integer> current, int total, Context context) {
        var candidates = context.getCandidates();
        var target = context.getTarget();
        var nCandidates = context.getCandidatesSize();

        if (total == target) {
            context.addToAnds(new ArrayList<>(current));
            return;
        }

        if ((i >= nCandidates) || (total > target)) {
            return;
        }

        current.add(candidates[i]);
        dfs(i, current, total + candidates[i], context);
        current.pop();
        dfs(i + 1, current, total, context);
    }

    private static final class Context {
        private final List<List<Integer>> ans = new ArrayList<>();
        private final int[] candidates;
        private final int target;

        public Context(int[] candidates, int target) {
            this.candidates = candidates;
            this.target = target;
        }

        public int getTarget() {
            return target;
        }

        public int getCandidatesSize() {
            return candidates.length;
        }

        public int[] getCandidates() {
            return candidates;
        }

        public List<List<Integer>> getAns() {
            return ans;
        }

        public void addToAnds(List<Integer> data) {
            ans.add(data);
        }
    }

}
